### PR TITLE
Matrix CWrappers

### PR DESCRIPTION
**Pull request by rajithv**

### CODE REVIEW

File: symengine/cwrapper.h [hunk 1]
```diff
@@ -5,6 +5,8 @@
 #include <stdlib.h>
 #include <gmp.h>
 
+#include "symengine/matrix.h"
```
Comment (isuruf) [inline symengine/cwrapper.h]: Don't include C++ headers in this C header. Take a look at `CVecBasic` to see how `vec_basic` is wrapped.

File: symengine/cwrapper.cpp [hunk 2]
```diff
@@ -563,6 +565,50 @@ size_t vecbasic_size(CVecBasic *self)
...
+{
+    return new CDenseMatrix;
+}
+
+void dense_matrix_free(CDenseMatrix *self)
+{
+    delete self;
+}
+
+void dense_matrix(CDenseMatrix *s)
```
Comment (abinashmeher999) [inline symengine/cwrapper.cpp]: This should be named something like `dense_matrix_init`.

File: symengine/cwrapper.cpp [hunk 3]
```diff
@@ -734,6 +780,12 @@ void ntheory_binomial(basic s, const basic a, unsigned long b)
     s->m = SymEngine::binomial(static_cast<const Integer &>(*(a->m)), b);
 }
 
+// Matrix Wrappers
+SymEngine::DenseMatrix* matrix_dense_matrix()
```
Comment (abinashmeher999) [inline symengine/cwrapper.cpp]: Do you plan to use this somewhere?

Comment (rajithv) [inline symengine/cwrapper.cpp]: That was actually something I wrote earlier, and had forgotten to remove. Now removed it.

File: symengine/cwrapper.cpp (hunk 2 above)
Comment (rajithv) [inline symengine/cwrapper.cpp]: noted

Comment (isuruf) [inline symengine/cwrapper.cpp]: @rajithv, you don't need this function right? You can use `dense_matrix_new()`, right?

**Comment by certik** [comment]
@rajithv please ping me once you fix the conflicts and want to review.

**Comment by certik** [comment]
(also remove the `[WIP]` once you want to review.)

**Comment by rajithv** [comment]
@certik will surely do. I've got to wrap the numpy-like functions and write the tests. Will ping you after completing those tasks.

File: symengine/cwrapper.cpp [hunk 4]
```diff
@@ -562,6 +565,194 @@ size_t vecbasic_size(CVecBasic *self)
...
+{
+    delete self;
+}
+
+void sparse_matrix_free(CSparseMatrix *self)
+{
+    delete self;
+}
+
+void dense_matrix_init(CDenseMatrix *s)
```
Comment (isuruf) [inline symengine/cwrapper.cpp]: What's the use of this function?

Comment (rajithv) [inline symengine/cwrapper.cpp]: wrapping `DenseMatrix::DenseMatrix(const DenseMatrix &x)` this constructor.
using it for: `SymEngine::DenseMatrix(SymEngine::DenseMatrix)` in the ruby wrapper

Comment (isuruf) [inline symengine/cwrapper.cpp]: `DenseMatrix::DenseMatrix(const DenseMatrix &x)` is used for copying. What's the difference with `dense_matrix_set`?

Comment (rajithv) [inline symengine/cwrapper.cpp]: Yes, I see it now. I will fix it. Thanks for pointing it out :)

File: symengine/cwrapper.cpp [hunk 5]
```diff
@@ -562,6 +569,258 @@ size_t vecbasic_size(CVecBasic *self)
...
+
+void sparse_matrix_init(CSparseMatrix *s)
+{
+    s->m = SymEngine::CSRMatrix();
+}
+
+void dense_matrix_rows_cols(CDenseMatrix *s, unsigned long int rows,
+                            unsigned long int cols)
+{
+    s->m = SymEngine::DenseMatrix(rows, cols);
```
Comment (isuruf) [inline symengine/cwrapper.cpp]: This shouldn't create a new Matrix and assign, but rather change `s`

File: symengine/cwrapper.cpp [hunk 6]
```diff
@@ -562,6 +569,258 @@ size_t vecbasic_size(CVecBasic *self)
...
+
+void sparse_matrix_rows_cols(CSparseMatrix *s, unsigned long int rows,
+                             unsigned long int cols)
+{
+    s->m = SymEngine::CSRMatrix(rows, cols);
+}
+
+void dense_matrix_set(CDenseMatrix *s, const CDenseMatrix *d)
+{
+    s->m = SymEngine::DenseMatrix(d->m);
```
Comment (isuruf) [inline symengine/cwrapper.cpp]: `s->m = d->m` is enough here

File: symengine/cwrapper.cpp [hunk 7]
```diff
@@ -562,6 +569,258 @@ size_t vecbasic_size(CVecBasic *self)
...
+    mat->m.set(r, c, s->m);
+}
+
+void dense_matrix_det(basic s, const CDenseMatrix *mat)
+{
+    s->m = mat->m.det();
+}
+void dense_matrix_inv(CDenseMatrix *s, const CDenseMatrix *mat)
+{
+    s->m = SymEngine::DenseMatrix();
```
Comment (isuruf) [inline symengine/cwrapper.cpp]: This line is unnecessary

File: symengine/cwrapper.cpp [hunk 8]
```diff
@@ -562,6 +569,258 @@ size_t vecbasic_size(CVecBasic *self)
...
+    dense_matrix_rows_cols(s, mat->m.nrows(), mat->m.ncols());
+    mat->m.inv(s->m);
+}
+void dense_matrix_transpose(CDenseMatrix *s, const CDenseMatrix *mat)
+{
+    s->m = SymEngine::DenseMatrix();
```
Comment (isuruf) [inline symengine/cwrapper.cpp]: Same here

**Comment by isuruf** [comment]
`s->m = SymEngine::DenseMatrix();` at the beginning of most functions is unnecessary

File: symengine/tests/cwrapper/test_cwrapper.c [hunk 9]
```diff
@@ -1013,7 +1118,8 @@ int main(int argc, char *argv[])
 #endif // HAVE_SYMENGINE_MPFR
 #ifdef HAVE_SYMENGINE_MPC
     test_complex_mpc();
-#endif // HAVE_SYMENGINE_MPC
-
+#endif // HAVE_SYMENGINE_MPC*/
+    symengine_print_stack_on_segfault();
+    test_matrix();
```
Comment (isuruf) [inline symengine/tests/cwrapper/test_cwrapper.c]: You can move matrix tests to a new test file if you want.

Comment (rajithv) [inline symengine/tests/cwrapper/test_cwrapper.c]: I'll uncomment the lines, once all the tests are done.
Commented them out, so running the tests would be easier locally.

File: symengine/matrix.h [hunk 10]
```diff
@@ -87,6 +87,9 @@ class DenseMatrix : public MatrixBase
     // Should implement all the virtual methods from MatrixBase
     // and throw an exception if a method is not applicable.
 
+    // Resize
+    virtual void resize(unsigned i, unsigned j);
+
```
Comment (isuruf) [inline symengine/matrix.h]: Can you remove `virtual` from this and move it above Line 87?

**Comment by rajithv** [comment]
@isuruf @abinashmeher999 @certik ready for review

File: symengine/cwrapper.cpp [hunk 11]
```diff
@@ -562,6 +569,252 @@ size_t vecbasic_size(CVecBasic *self)
...
+
+void sparse_matrix_rows_cols(CSparseMatrix *s, unsigned long int rows,
+                             unsigned long int cols)
+{
+    s->m = SymEngine::CSRMatrix(rows, cols);
+}
+
+void dense_matrix_set(CDenseMatrix *s, const CDenseMatrix *d)
+{
+    s->m = d->m;
```
Comment (isuruf) [inline symengine/cwrapper.cpp]: Can you add a test for this?

File: symengine/tests/cwrapper/test_cwrapper.c [hunk 12]
```diff
@@ -985,6 +985,262 @@ void test_eval()
...
+    expected = "[-2, 0, 0, 0]\n[0, 3, 0, 0]\n[0, 0, 3, 0]\n[0, 0, 0, -4]\n";
+    SYMENGINE_C_ASSERT(strcmp(result, expected) == 0);
+    
+    //det
+    dense_matrix_det(i1, D);
+    SYMENGINE_C_ASSERT(integer_get_ui(i1) == 72);
+    
+    // eye
+    dense_matrix_eye(D, 3, 4, 1);
+    result = dense_matrix_str(D);
```
Comment (isuruf) [inline symengine/tests/cwrapper/test_cwrapper.c]: `result` should be deallocated every time you use `dense_matrix_str`

File: symengine/tests/cwrapper/test_cwrapper.c [hunk 13]
```diff
@@ -1157,45 +1163,91 @@ void test_matrix()
...
+    basic_str_free(result);
     
     result = dense_matrix_str(D);
     expected = "[4, 0]\n[0, -1/4]\n";
     SYMENGINE_C_ASSERT(strcmp(result, expected) == 0);
+    basic_str_free(result);
+    
+    // set matrix
+    dense_matrix_set(C, D);
+    SYMENGINE_C_ASSERT(dense_matrix_eq(D, C) == 1);
```
Comment (isuruf) [inline symengine/tests/cwrapper/test_cwrapper.c]: Can you change `D` after this line and check that this doesn't affect `C` ?

**Comment by isuruf** [comment]
+1 to merge when tests pass

**Comment by certik** [comment]
Thanks @rajithv for wrapping these!

**Pull request merged**